File: backend/mr/src/main/java/io/featurehub/mr/resources/AuthResource.java
package io.featurehub.mr.resources;

import cd.connect.app.config.ConfigKey;
import io.featurehub.db.api.AuthenticationApi;
import io.featurehub.db.api.FillOpts;
import io.featurehub.db.api.Opts;
import io.featurehub.db.api.PersonApi;
import io.featurehub.mr.api.AuthServiceDelegate;
import io.featurehub.mr.auth.AuthManagerService;
import io.featurehub.mr.auth.AuthenticationRepository;
import io.featurehub.mr.model.PasswordReset;
import io.featurehub.mr.model.PasswordUpdate;
import io.featurehub.mr.model.Person;
import io.featurehub.mr.model.PersonRegistrationDetails;
import io.featurehub.mr.model.ProviderRedirect;
import io.featurehub.mr.model.RegistrationUrl;
import io.featurehub.mr.model.TokenizedPerson;
import io.featurehub.mr.model.UserCredentials;
import io.featurehub.mr.resources.auth.AuthProvider;
import org.glassfish.hk2.api.IterableProvider;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.ws.rs.BadRequestException;
import javax.ws.rs.ForbiddenException;
import javax.ws.rs.NotFoundException;
import javax.ws.rs.core.SecurityContext;
import java.util.ArrayList;
import java.util.List;

public class AuthResource implements AuthServiceDelegate {
  private static final Logger log = LoggerFactory.getLogger(AuthResource.class);
  private final AuthenticationApi authenticationApi;
  private final AuthManagerService authManager;
  private final PersonApi personApi;
  private final AuthenticationRepository authRepository;
  private final List<AuthProvider> authProviders = new ArrayList<>();

  @ConfigKey("auth.disable-login")
  protected Boolean loginDisabled = Boolean.FALSE;

  @Inject
  public AuthResource(AuthenticationApi authenticationApi, AuthManagerService authManager, PersonApi personApi, AuthenticationRepository authRepository, IterableProvider<AuthProvider> authProviders) {
    this.authenticationApi = authenticationApi;
    this.authManager = authManager;
    this.personApi = personApi;
    this.authRepository = authRepository;
    if (authProviders != null) {
      authProviders.forEach(this.authProviders::add);
    }
  }

  @Override
  public Person changePassword(String id, PasswordUpdate passwordUpdate, SecurityContext securityContext) {
    Person personByToken = authManager.from(securityContext);

    // yourself or a superuser can change your password. This allows a superuser to change the password immediately
    // after reset without having to go to any further trouble.
    if (personByToken.getId().getId().equals(id) || authManager.isOrgAdmin(personByToken)) {
      Person newPerson = authenticationApi.changePassword(id, passwordUpdate.getOldPassword(), passwordUpdate.getNewPassword());

      if (newPerson == null) {
        throw new BadRequestException("Old password does not match.");
      }

      return newPerson;
    }

    throw new ForbiddenException();
  }

  /**
   * We have to do this at request time, because some urls are time sensitive (e.g. oauth ones with the state
   * parameter).
   *
   * @param provider
   * @return
   */
  @Override
  public ProviderRedirect getLoginUrlForProvider(String provider) {
    for(AuthProvider ap : authProviders) {
      if (ap.getProviders().contains(provider)) {
        return  new ProviderRedirect().redirectUrl(ap.requestRedirectUrl(provider));
      }
    }

    throw new NotFoundException();
  }

  @Override
  public TokenizedPerson login(UserCredentials userCredentials) {
    // if access via this API is forbidden (for example only GUI based OAuth or SAML login is allowed)
    // then fail requests automatically
    if (loginDisabled) {
      throw new ForbiddenException();
    }

    // can't try and login with a null or empty password
    if (userCredentials.getPassword() == null || userCredentials.getPassword().trim().length() == 0) {
      throw new ForbiddenException();
    }

    Person login = authenticationApi.login(userCredentials.getEmail(), userCredentials.getPassword());

    if (login == null) {
      throw new NotFoundException();
    }

    return new TokenizedPerson().accessToken(authRepository.put(login)).person(login);
  }

  @Override
  public void logout(SecurityContext securityContext) {
    authRepository.invalidate(securityContext);
  }

  @Override
  public Person personByToken(String token) {
    Person personByToken = authenticationApi.getPersonByToken(token);

    if (personByToken == null) {
      throw new NotFoundException("No person by that token");
    }

    return personByToken;
  }

  @Override
  public TokenizedPerson registerPerson(PersonRegistrationDetails personRegistrationDetails) {

    //check user found by token and token hasn't expired

    Person person = personApi.getByToken(personRegistrationDetails.getRegistrationToken(), Opts.opts(FillOpts.Groups));

    if (person == null) {
      throw new NotFoundException("Person already registered using token");
    }

    if (!person.getEmail().toLowerCase().equals(personRegistrationDetails.getEmail().toLowerCase())) {
      log.info("db user email `{}` does not match passed email `{}`", person.getEmail(), personRegistrationDetails.getEmail());
      //registration token doesn't belong to provided email
      throw new BadRequestException();
    }

    if (!personRegistrationDetails.getPassword().equals(personRegistrationDetails.getConfirmPassword())) {
      //passwords don't match
      throw new BadRequestException();
    }

    Person newPerson = authenticationApi.register(personRegistrationDetails.getName(),
      personRegistrationDetails.getEmail(),
      personRegistrationDetails.getPassword());

    if (newPerson == null) {
      throw new NotFoundException("Cannot find person to register");
    }

    return new TokenizedPerson().accessToken(authRepository.put(person.copy())).person(newPerson);
  }

  @Override
  public TokenizedPerson replaceTempPassword(String id, PasswordReset passwordReset, SecurityContext context) {
    Person person = authManager.from(context);

    if (Boolean.TRUE.equals(person.getPasswordRequiresReset())) {
      if (person.getId().getId().equals(id)) { // its me
        Person newPerson = authenticationApi.replaceTemporaryPassword(id, passwordReset.getPassword());
        authRepository.invalidate(context);
        return new TokenizedPerson().accessToken(authRepository.put(newPerson)).person(newPerson);
      }
    }

    throw new ForbiddenException();
  }

  @Override
  public RegistrationUrl resetExpiredToken(String email, SecurityContext context) {
    Person person = authManager.from(context);

    if (authManager.isAnyAdmin(person)) {
      String token = authenticationApi.resetExpiredRegistrationToken(email);

      return new RegistrationUrl().registrationUrl(token);
    }

    throw new ForbiddenException();
  }

  @Override
  public Person resetPassword(String id, PasswordReset passwordReset, SecurityContext context) {
    if (authManager.isAnyAdmin(authManager.from(context))) {
      Person person = authenticationApi.resetPassword(id, passwordReset.getPassword(),
        authManager.from(context).getId().getId(),
        Boolean.TRUE.equals(passwordReset.getReactivate()));

      if (person == null) {
        throw new NotFoundException();
      }

      return person;
    }

    throw new ForbiddenException();
  }
}

File: backend/mr/src/main/kotlin/io/featurehub/mr/resources/AuthResource.kt
package io.featurehub.mr.resources

import cd.connect.app.config.ConfigKey
import cd.connect.app.config.DeclaredConfigResolver
import io.featurehub.db.api.AuthenticationApi
import io.featurehub.db.api.FillOpts
import io.featurehub.db.api.Opts
import io.featurehub.db.api.PersonApi
import io.featurehub.db.password.PasswordPolicy
import io.featurehub.db.password.PasswordPolicyViolationException
import io.featurehub.mr.api.AuthServiceDelegate
import io.featurehub.mr.auth.AuthManagerService
import io.featurehub.mr.auth.AuthenticationRepository
import io.featurehub.mr.model.*
import io.featurehub.web.security.oauth.AuthProviderCollection
import jakarta.annotation.security.PermitAll
import jakarta.inject.Inject
import jakarta.ws.rs.BadRequestException
import jakarta.ws.rs.ForbiddenException
import jakarta.ws.rs.NotFoundException
import jakarta.ws.rs.WebApplicationException
import jakarta.ws.rs.core.Response
import jakarta.ws.rs.core.SecurityContext
import org.slf4j.LoggerFactory
import java.util.*
import kotlin.String

class AuthResource @Inject constructor(
  private val authenticationApi: AuthenticationApi,
  private val authManager: AuthManagerService, private val personApi: PersonApi,
  private val authRepository: AuthenticationRepository, private val authProviderCollection: AuthProviderCollection,
  private val passwordPolicy: PasswordPolicy
) : AuthServiceDelegate {
  private val log = LoggerFactory.getLogger(AuthResource::class.java)

  @ConfigKey("auth.disable-login")
  private var loginDisabled:Boolean? = false

  init {
    DeclaredConfigResolver.resolve(this)
  }

  /**
   * Rejects a password that fails the complexity policy with a 400 carrying every rule it violated,
   * so the (localised) admin console can render its own guidance.
   *
   * Only ever called on a path that *sets* a password. Login and the old-password half of a change
   * request deliberately do not go through here - stored passwords predate this policy and their
   * owners must still be able to authenticate. See PasswordPolicy.
   */
  private fun enforcePasswordPolicy(password: String?) {
    try {
      passwordPolicy.validate(password)
    } catch (e: PasswordPolicyViolationException) {
      throw WebApplicationException(
        Response.status(Response.Status.BAD_REQUEST)
          .entity(
            PasswordPolicyViolationReport()
              .violatedRules(e.violations.map { PasswordPolicyRule.valueOf(it.name) })
              .minLength(e.minLength)
              .maxLength(e.maxLength)
          )
          .build()
      )
    }
  }

  override fun changePassword(id: UUID, passwordUpdate: PasswordUpdate, securityContext: SecurityContext?): Person {
    val personByToken = authManager.from(securityContext)

    // yourself or a superuser can change your password. This allows a superuser to change the password immediately
    // after reset without having to go to any further trouble.
    if (personByToken.id!!.id == id || authManager.isOrgAdmin(personByToken)) {
      // only the new password is policed - the old one is an existing credential being verified
      enforcePasswordPolicy(passwordUpdate.newPassword)

      return authenticationApi.changePassword(id, passwordUpdate.oldPassword, passwordUpdate.newPassword)
        ?: throw BadRequestException("Old password does not match.")
    }
    throw ForbiddenException()
  }

  /**
   * We have to do this at request time, because some urls are time sensitive (e.g. oauth ones with the state
   * parameter).
   *
   * @param provider
   * @return
   */
  override fun getLoginUrlForProvider(provider: String): ProviderRedirect {
    val authProviderSource = authProviderCollection.find(provider)
    if (authProviderSource != null) {
      return ProviderRedirect().redirectUrl(authProviderSource.redirectUrl!!)
    }
    throw NotFoundException()
  }

  override fun login(userCredentials: UserCredentials): TokenizedPerson {
    // if access via this API is forbidden (for example only GUI based OAuth or SAML login is allowed)
    // then fail requests automatically
    if (loginDisabled == true) {
      throw ForbiddenException()
    }

    // can't try and login with a null or empty password
    if (userCredentials.password.trim { it <= ' ' }.isEmpty()) {
      throw ForbiddenException()
    }

    val login = authenticationApi.login(userCredentials.email, userCredentials.password)
      ?: throw NotFoundException()
    return TokenizedPerson().accessToken(authRepository.put(login)).person(login)
  }

  @PermitAll
  override fun logout(securityContext: SecurityContext?) {
    authRepository.invalidate(securityContext)
  }

  override fun personByToken(token: String): Person {
    return authenticationApi.getPersonByToken(token)
      ?: throw NotFoundException("No person by that token")
  }

  override fun registerPerson(personRegistrationDetails: PersonRegistrationDetails): TokenizedPerson {

    //check user found by token and token hasn't expired
    val person = personApi.getByToken(personRegistrationDetails.registrationToken, Opts.opts(FillOpts.Groups))
      ?: throw NotFoundException("Person already registered using token")

    if (person.email == null || !person.email.equals(
        personRegistrationDetails.email.lowercase(Locale.getDefault()),
        ignoreCase = true
      )
    ) {
      log.info("db user email `{}` does not match passed email `{}`", person.email, personRegistrationDetails.email)
      throw BadRequestException()
    }

    if (personRegistrationDetails.password != personRegistrationDetails.confirmPassword) {
      //passwords don't match
      throw BadRequestException()
    }

    enforcePasswordPolicy(personRegistrationDetails.password)

    val newPerson = authenticationApi.register(
      personRegistrationDetails.name,
      personRegistrationDetails.email,
      personRegistrationDetails.password, null
    ) ?: throw NotFoundException("Cannot find person to register")

    return TokenizedPerson().accessToken(authRepository.put(person.copy())).person(newPerson)
  }

  override fun replaceTempPassword(
    id: UUID,
    passwordReset: PasswordReset,
    context: SecurityContext?
  ): TokenizedPerson {
    val person = authManager.from(context)
    if (true == person.passwordRequiresReset) {
      if (person.id!!.id == id) { // its me
        enforcePasswordPolicy(passwordReset.password)

        val newPerson = authenticationApi.replaceTemporaryPassword(id, passwordReset.password)
        authRepository.invalidate(context)
        return TokenizedPerson().accessToken(authRepository.put(newPerson)).person(newPerson)
      }
    }
    throw ForbiddenException()
  }

  override fun resetExpiredToken(email: String, context: SecurityContext?): RegistrationUrl {
    val person = authManager.from(context)
    if (authManager.isAnyAdmin(person)) {
      val token = authenticationApi.resetExpiredRegistrationToken(email)
      return RegistrationUrl()
        .personId(person.id!!.id)
        .registrationUrl(token)
        .token(token)
    }
    throw ForbiddenException()
  }

  override fun resetPassword(id: UUID, passwordReset: PasswordReset, context: SecurityContext?): Person {
    if (authManager.isAnyAdmin(authManager.from(context))) {
      // after the admin check, so an unauthorised caller gets a bare 403 and learns nothing about the policy
      enforcePasswordPolicy(passwordReset.password)

      return authenticationApi.resetPassword(
        id, passwordReset.password,
        authManager.from(context).id!!.id, true == passwordReset.reactivate
      )
        ?: throw NotFoundException()
    }
    throw ForbiddenException()
  }
}
